spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/UDTValue.scala

```diff
@@ -6,24 +6,22 @@ import scala.reflect.runtime.universe._
 import com.datastax.driver.core.{UDTValue => DriverUDTValue}
 import com.datastax.spark.connector.types.NullableTypeConverter
 
-final case class UDTValue(columnNames: IndexedSeq[String], columnValues: IndexedSeq[AnyRef])
+final case class UDTValue(metaData: CassandraRowMetadata, columnValues: IndexedSeq[AnyRef])
```

Comment:
@jtgrabowski for the sake of preserving API compatibility, if needed, I can suggest the following change:

````suggestion
final case class UDTValue(columnNames: IndexedSeq[String], columnValues: IndexedSeq[AnyRef])
  extends ScalaGettableData {

  // caching CassandraRowMetadata instance
  @volatile private var metaData_ : Option[CassandraRowMetadata] = None

  //alternative constructor
  def this(metaData: CassandraRowMetadata, columnValues: IndexedSeq[AnyRef]) = {
    this(metaData.columnNames, columnValues)
    this.metaData_ = Some(metaData)
  }

...

  override def metaData: CassandraRowMetadata = {
    metaData_ match {
      case Some(x) => x
      // I think it's ok to assign without double checked locking here:
      // 1. earlier each call of metadata() method resulted in new CassandraRowMetadata instance => returning several different instances should be fine
      // 2. if the caller did not supply CassandraRowMetadata, they are not concerned about performance
      case None => val x = CassandraRowMetadata.fromColumnNames(columnNames); metaData_ = Some(x); x
    }
  }
}

object UDTValue {

...
  // alternative apply method
  def apply(metaData: CassandraRowMetadata, columnValues: IndexedSeq[AnyRef]): UDTValue =
    new UDTValue(metaData, columnValues)
}
````

Reply:
It's a good idea, let's stay with the original signature (```final case class UDTValue(columnNames: IndexedSeq[String], columnValues: IndexedSeq[AnyRef])```).
Instead of a volatile var use a lazy val and we should be good to go.

Reply:
Hi @jtgrabowski ! I've update the pull request with the corresponding changes - both this one and b2.5 one either.
I've run test and it:test  - they were fine. My internal perf test was ok either.